staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go

```diff
@@ -255,18 +255,13 @@ func (e DirectEncoder) Encode(obj runtime.Object, stream io.Writer) error {
 	return err
 }
 
-// DirectDecoder clears the group version kind of a deserialized object.
+// DirectDecoder does not do conversion.
 type DirectDecoder struct {
 	runtime.Decoder
 }
 
-// Decode does not do conversion. It removes the gvk during deserialization.
+// Decode does not do conversion.
 func (d DirectDecoder) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
 	obj, gvk, err := d.Decoder.Decode(data, defaults, into)
-	if obj != nil {
```

Comment:
When doing a decoding from bytes, we definitively know the GVK as specified by the byte stream. TypeMeta isn't set when constructing an object from scratch since we don't know the type it will eventually be serialized as (usually), but when decoding from bytes in a DirectDecoder (it doesn't do conversion) we do know the value.

@kubernetes/sig-api-machinery-pr-reviews @smarterclayton @lavalamp

Reply:
I don't know that this would definitely break anyone.  It's only there because of some historical confusion.

An encoder will still stomp the GVK no matter what, so I have trouble imagining how this is bad.

Reply:
We used to have the rule that this stuff was set on the wire and blank in memory (because it's implied by the go type and this way you can't get it wrong). I don't know if this rule is still necessary, but what is the new rule that is being proposed? I think there should be some invariant about this stuff. And I would hate to be in a situation where I have to set it everywhere I make an object, as in this PR; it's super repetitive.

Reply:
> We used to have the rule that this stuff was set on the wire and blank in memory (because it's implied by the go type and this way you can't get it wrong). I don't know if this rule is still necessary, but what is the new rule that is being proposed? I think there should be some invariant about this stuff. And I would hate to be in a situation where I have to set it everywhere I make an object, as in this PR; it's super repetitive.

The rule was about the information available.  A golang type does not contain enough information to know how it is going to be serialized.  The changes made in this pull during construction are safe because we *do* know how they are going to be serialized.

If the information is known, it can be present.  This means that consuming the value via decoding bytes (client or reading a file for instance) it is known.  In many cases you can know when you're constructing, but not in all.  For instance, most openshift objects don't know.  No internal types know.

Reply:
> And I would hate to be in a situation where I have to set it everywhere I make an object, as in this PR; it's super repetitive.

You only have to set it when you're trying to simulate decoding from bytes.  That's why so few places (yeah, it looks like a bunch, but it's really pretty contained) cared.

Reply:
I think I'm fine with this. But I would like some invariant, e.g. "version/kind is set when _____ and not set when ____" to explain to people.